File: leetcode/FindPivotIndex.py
from typing import List

class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        """
        Finds the pivot index in an array.
        A pivot index is the index where the sum of all the numbers strictly to the left
        of the index is equal to the sum of all the numbers strictly to the index's right.
        If no such index exists, return -1.
        """
        # Initialize 'left_sum' to 0, representing the sum of elements to the left of the current index.
        # Initialize 'right_sum' to the total sum of all elements, representing the sum of elements
        # to the right of the current index (before subtracting the current element).
        left_sum, right_sum = 0, sum(nums)
        
        for i, num in enumerate(nums):
            # Subtract the current number from 'right_sum' to get the sum of elements strictly to its right.
            right_sum -= num
            
            # If 'left_sum' equals 'right_sum', we found the pivot index.
            if left_sum == right_sum:
                return i
            
            # Add the current number to 'left_sum' for the next iteration.
            # This prepares 'left_sum' to be the sum of elements strictly to the left of the next element.
            left_sum += num
            
        # If the loop completes without finding a pivot, return -1.
        return -1

File: leetcode/easy/FindPivotIndex.java
import java.util.Arrays;

/**
 * Solution class for finding the pivot index in an array.
 */
class Solution {
    /**
     * Given an array of integers nums, this method finds the pivot index.
     * The pivot index is where the sum of numbers to the left of the index
     * equals the sum of numbers to the right of the index.
     *
     * @param nums The input array of integers.
     * @return The leftmost pivot index, or -1 if no such index exists.
     */
    public int pivotIndex(int[] nums) {
        // First, calculate the total sum of all numbers in the array.
        int totalSum = Arrays.stream(nums).sum();

        // Initialize leftSum, which will store the sum of numbers to the left of the current index.
        int leftSum = 0;

        // Iterate through each element of the array to check if it's a pivot index.
        for (int i = 0; i < nums.length; i++) {
            // The right sum is the total sum minus the left sum and the current element's value.
            int rightSum = totalSum - leftSum - nums[i];

            // Check if the sum of numbers to the left equals the sum of numbers to the right.
            if (leftSum == rightSum) {
                // If they are equal, we've found the pivot index.
                return i;
            }

            // If it's not a pivot, add the current number to the leftSum for the next iteration.
            leftSum += nums[i];
        }

        // If the loop completes without finding a pivot index, it means no pivot exists.
        return -1;
    }
}
